### PR TITLE
Migrate collections to PHP 8 syntax

### PR DESCRIPTION
**Pull request by greg0ire**

### CODE REVIEW

File: lib/Doctrine/ORM/LazyCriteriaCollection.php
Comment (greg0ire) [inline lib/Doctrine/ORM/LazyCriteriaCollection.php]: [Maybe we should switch to Psalm v5 because of this](https://github.com/vimeo/psalm/issues/8229)

File: lib/Doctrine/ORM/LazyCriteriaCollection.php
Comment (derrabus) [inline lib/Doctrine/ORM/LazyCriteriaCollection.php:83]: Can we add `Selectable` to the native return type?

```suggestion
    public function matching(Criteria $criteria): ReadableCollection&Selectable
```

File: lib/Doctrine/ORM/PersistentCollection.php
Comment (derrabus) [inline lib/Doctrine/ORM/PersistentCollection.php]:
```suggestion
    private array|null $association = null;
```

File: lib/Doctrine/ORM/LazyCriteriaCollection.php
Comment (greg0ire) [reply lib/Doctrine/ORM/LazyCriteriaCollection.php:83]: I haven't found how to do so, see https://github.com/doctrine/orm/pull/10084#discussion_r984973270

Comment (derrabus) [reply lib/Doctrine/ORM/LazyCriteriaCollection.php:83]: So we're not adding the correct native return type because it'll make Psalm sad? Shall we rather ignore the error?

Comment (greg0ire) [reply lib/Doctrine/ORM/LazyCriteriaCollection.php:83]: It haven't found how to ignore it: it's a crash, not a simple error.

Comment (derrabus) [reply lib/Doctrine/ORM/LazyCriteriaCollection.php:83]: 😭

Comment (greg0ire) [reply lib/Doctrine/ORM/LazyCriteriaCollection.php:83]: @derrabus I've discussed this here: https://symfony-devs.slack.com/archives/C8SFXTD2M/p1665220038738889

I see 4 options:
1. wait for Psalm 5 before merging this;
2. merge without the intersection type, wait for a merge up between Psalm `4.x` and `master` and switch all branches to Psalm `master` branch, and add the type;
3. merge without the intersection type, wait for Psalm 5 to be released and add the type then
4. switch to Psalm `master` branch now and add the type.

I don't see switching only 3.0.x to another version of Psalm as a reasonable option.

2 and 3 have my preference, what about you?

Comment (derrabus) [reply lib/Doctrine/ORM/LazyCriteriaCollection.php:83]: 2 sounds good.

**Comment by greg0ire** [comment]
Blocked by https://github.com/doctrine/orm/pull/10116 and a merge up

File: lib/Doctrine/ORM/LazyCriteriaCollection.php
Comment (greg0ire) [reply lib/Doctrine/ORM/LazyCriteriaCollection.php:83]: Extracted that into a new issue: https://github.com/doctrine/orm/issues/10118

**Review by derrabus — APPROVED** [review]

**Pull request merged**
